File: ios/Godstone/Sources/GodstoneMesh/PeerTrustModels.swift
import Foundation
import CryptoKit

/// Peer trust level with explicit persistence codes (ADR-003, Phase C8.2A).
///
/// Explicit Int32 codes are used instead of enum ordinals to prevent database
/// reinterpretation across schema evolutions:
/// - tofuPinned = 1
/// - userVerified = 2
/// - revoked = 3
public enum PeerTrustLevel: Int32, Sendable, Equatable {
    case tofuPinned = 1
    case userVerified = 2
    case revoked = 3

    public static func fromPersistedCode(_ code: Int32) -> PeerTrustLevel? {
        PeerTrustLevel(rawValue: code)
    }
}

/// Effective trust state computed from `PeerIdentityRecord` (ADR-003, Phase C8.2A).
///
/// Precedence rule:
/// 1. revoked -> revoked
/// 2. pendingGeneration != nil -> keyChangedQuarantined
/// 3. userVerified -> activeUserVerified
/// 4. tofuPinned -> activeTofu
public enum EffectivePeerTrustState: Sendable, Equatable {
    case activeTofu
    case activeUserVerified
    case keyChangedQuarantined
    case revoked
}

/// Authoritative cryptographic peer identity record (ADR-003, Phase C8.2A).
///
/// Contains only cryptographic trust authority fields. Contact names, metadata,
/// message-store state, and routing hints are strictly prohibited in this model.
public struct PeerIdentityRecord: Sendable, Equatable {
    public let nodeId: Data
    public let signingPublicKey: Data
    public let acceptedStaticDhPublicKey: Data
    public let acceptedGeneration: UInt32
    public let trustLevel: PeerTrustLevel
    public let pendingStaticDhPublicKey: Data?
    public let pendingGeneration: UInt32?

    public init(
        nodeId: Data,
        signingPublicKey: Data,
        acceptedStaticDhPublicKey: Data,
        acceptedGeneration: UInt32,
        trustLevel: PeerTrustLevel,
        pendingStaticDhPublicKey: Data? = nil,
        pendingGeneration: UInt32? = nil
    ) {
        self.nodeId = nodeId
        self.signingPublicKey = signingPublicKey
        self.acceptedStaticDhPublicKey = acceptedStaticDhPublicKey
        self.acceptedGeneration = acceptedGeneration
        self.trustLevel = trustLevel
        self.pendingStaticDhPublicKey = pendingStaticDhPublicKey
        self.pendingGeneration = pendingGeneration
    }

    public var effectiveState: EffectivePeerTrustState {
        if trustLevel == .revoked {
            return .revoked
        } else if pendingGeneration != nil {
            return .keyChangedQuarantined
        } else if trustLevel == .userVerified {
            return .activeUserVerified
        } else {
            return .activeTofu
        }
    }
}

/// Corruption reasons for durable `PeerIdentityRecord` validation (ADR-003, Phase C8.2A).
public enum PeerRecordCorruptionReason: Sendable, Equatable {
    case invalidNodeIdLength
    case invalidSigningKeyLength
    case invalidAcceptedStaticKeyLength
    case acceptedGenerationOutOfRange
    case nodeIdSigningKeyMismatch
    case pendingCouplingViolation
    case invalidPendingStaticKeyLength
    case pendingGenerationOutOfRange
    case pendingNotNewer
    case pendingStaticEqualsAccepted
    case revokedWithPending
}

/// Validation result taxonomy for durable `PeerIdentityRecord` (ADR-003, Phase C8.2A).
public enum PeerIdentityRecordValidationResult: Sendable, Equatable {
    case valid
    case corrupt(PeerRecordCorruptionReason)
}

/// Pure validator enforcing the 11 durable invariants of `PeerIdentityRecord` (ADR-003, Phase C8.2A).
public enum PeerIdentityRecordValidator {
    public static let nodeIdLength = 16
    public static let signingKeyLength = 32
    public static let staticKeyLength = 32

    public static func validate(record: PeerIdentityRecord) -> PeerIdentityRecordValidationResult {
        // R1: nodeId exactly 16 bytes
        guard record.nodeId.count == nodeIdLength else {
            return .corrupt(.invalidNodeIdLength)
        }

        // R2: signingPublicKey exactly 32 bytes
        guard record.signingPublicKey.count == signingKeyLength else {
            return .corrupt(.invalidSigningKeyLength)
        }

        // R3: acceptedStaticDhPublicKey exactly 32 bytes
        guard record.acceptedStaticDhPublicKey.count == staticKeyLength else {
            return .corrupt(.invalidAcceptedStaticKeyLength)
        }

        // R5: nodeId == BLAKE2s-128(signingPublicKey)
        let expectedNodeId = IdentityBindingV1.deriveNodeId(signingPublicKey: record.signingPublicKey)
        guard record.nodeId == expectedNodeId else {
            return .corrupt(.nodeIdSigningKeyMismatch)
        }

        // R6: pendingStaticDhPublicKey and pendingGeneration are BOTH nil or BOTH non-nil
        let pendStatic = record.pendingStaticDhPublicKey
        let pendGen = record.pendingGeneration
        guard (pendStatic == nil) == (pendGen == nil) else {
            return .corrupt(.pendingCouplingViolation)
        }

        // Pending field invariants
        if let pendStatic = pendStatic, let pendGen = pendGen {
            // R11: if trustLevel == revoked, pending fields MUST both be nil
            guard record.trustLevel != .revoked else {
                return .corrupt(.revokedWithPending)
            }

            // R7: pending static exactly 32 bytes
            guard pendStatic.count == staticKeyLength else {
                return .corrupt(.invalidPendingStaticKeyLength)
            }

            // R8: pendingGeneration > acceptedGeneration
            guard pendGen > record.acceptedGeneration else {
                return .corrupt(.pendingNotNewer)
            }

            // R10: pendingStaticDhPublicKey != acceptedStaticDhPublicKey
            guard pendStatic != record.acceptedStaticDhPublicKey else {
                return .corrupt(.pendingStaticEqualsAccepted)
            }
        }

        return .valid
    }
}

/// Rejection reason taxonomy for peer trust evaluation (ADR-003, Phase C8.2A).
public enum PeerTrustRejectReason: Sendable, Equatable {
    case rollback
    case sameGenerationConflict
    case pendingGenerationConflict
    case staleRelativeToPending
    case noncanonicalGenerationAdvance
    case nodeIdSigningKeyCollision
    case revoked
}

/// Pure decision plan emitted by `PeerTrustEngine` (ADR-003, Phase C8.2A).
public enum TrustPlan: Sendable, Equatable {
    case acceptExisting
    case insertFirstSeen
    case setInitialPendingCandidate
    case advancePendingCandidate
    case keepQuarantined
    case reject(PeerTrustRejectReason)
}

/// Read-only view of a verified active peer identity (ADR-003, Phase C8.2A).
public struct VerifiedPeerIdentity: Sendable, Equatable {
    public let nodeId: Data
    public let signingPublicKey: Data
    public let acceptedStaticDhPublicKey: Data
    public let acceptedGeneration: UInt32
    public let trustLevel: PeerTrustLevel

    fileprivate init(
        nodeId: Data,
        signingPublicKey: Data,
        acceptedStaticDhPublicKey: Data,
        acceptedGeneration: UInt32,
        trustLevel: PeerTrustLevel
    ) {
        self.nodeId = nodeId
        self.signingPublicKey = signingPublicKey
        self.acceptedStaticDhPublicKey = acceptedStaticDhPublicKey
        self.acceptedGeneration = acceptedGeneration
        self.trustLevel = trustLevel
    }

    public static func fromRecord(_ record: PeerIdentityRecord) -> VerifiedPeerIdentity? {
        guard case .valid = PeerIdentityRecordValidator.validate(record: record) else {
            return nil
        }
        guard record.trustLevel != .revoked, record.pendingGeneration == nil else {
            return nil
        }
        return VerifiedPeerIdentity(
            nodeId: record.nodeId,
            signingPublicKey: record.signingPublicKey,
            acceptedStaticDhPublicKey: record.acceptedStaticDhPublicKey,
            acceptedGeneration: record.acceptedGeneration,
            trustLevel: record.trustLevel
        )
    }
}

/// Read-only view of a quarantined peer identity with a pending rotation candidate (ADR-003, Phase C8.2A).
public struct PendingPeerIdentity: Sendable, Equatable {
    public let nodeId: Data
    public let signingPublicKey: Data
    public let acceptedStaticDhPublicKey: Data
    public let acceptedGeneration: UInt32
    public let trustLevel: PeerTrustLevel
    public let pendingStaticDhPublicKey: Data
    public let pendingGeneration: UInt32

    fileprivate init(
        nodeId: Data,
        signingPublicKey: Data,
        acceptedStaticDhPublicKey: Data,
        acceptedGeneration: UInt32,
        trustLevel: PeerTrustLevel,
        pendingStaticDhPublicKey: Data,
        pendingGeneration: UInt32
    ) {
        self.nodeId = nodeId
        self.signingPublicKey = signingPublicKey
        self.acceptedStaticDhPublicKey = acceptedStaticDhPublicKey
        self.acceptedGeneration = acceptedGeneration
        self.trustLevel = trustLevel
        self.pendingStaticDhPublicKey = pendingStaticDhPublicKey
        self.pendingGeneration = pendingGeneration
    }

    public static func fromRecord(_ record: PeerIdentityRecord) -> PendingPeerIdentity? {
        guard case .valid = PeerIdentityRecordValidator.validate(record: record) else {
            return nil
        }
        guard record.trustLevel != .revoked,
              let pendStatic = record.pendingStaticDhPublicKey,
              let pendGen = record.pendingGeneration else {
            return nil
        }
        return PendingPeerIdentity(
            nodeId: record.nodeId,
            signingPublicKey: record.signingPublicKey,
            acceptedStaticDhPublicKey: record.acceptedStaticDhPublicKey,
            acceptedGeneration: record.acceptedGeneration,
            trustLevel: record.trustLevel,
            pendingStaticDhPublicKey: pendStatic,
            pendingGeneration: pendGen
        )
    }
}

File: android/mesh/src/main/java/io/godstone/mesh/identity/PeerTrustModels.kt
package io.godstone.mesh.identity

/**
 * Peer trust level with explicit persistence codes (ADR-003, Phase C8.2A).
 *
 * Explicit integer codes are used instead of enum ordinals to prevent database
 * reinterpretation across schema evolutions:
 * - TOFU_PINNED = 1
 * - USER_VERIFIED = 2
 * - REVOKED = 3
 */
enum class PeerTrustLevel(val persistedCode: Int) {
    TOFU_PINNED(1),
    USER_VERIFIED(2),
    REVOKED(3);

    companion object {
        fun fromPersistedCode(code: Int): PeerTrustLevel? = when (code) {
            1 -> TOFU_PINNED
            2 -> USER_VERIFIED
            3 -> REVOKED
            else -> null
        }
    }
}

/**
 * Effective trust state computed from [PeerIdentityRecord] (ADR-003, Phase C8.2A).
 *
 * Precedence rule:
 * 1. REVOKED -> REVOKED
 * 2. pendingGeneration != null -> KEY_CHANGED_QUARANTINED
 * 3. USER_VERIFIED -> ACTIVE_USER_VERIFIED
 * 4. TOFU_PINNED -> ACTIVE_TOFU
 */
internal enum class EffectivePeerTrustState {
    ACTIVE_TOFU,
    ACTIVE_USER_VERIFIED,
    KEY_CHANGED_QUARANTINED,
    REVOKED
}

/**
 * Authoritative cryptographic peer identity record (ADR-003, Phase C8.2A).
 *
 * Module-internal storage and repository model only. Contains only cryptographic
 * trust authority fields.
 */
internal class PeerIdentityRecord(
    nodeId: ByteArray,
    signingPublicKey: ByteArray,
    acceptedStaticDhPublicKey: ByteArray,
    val acceptedGeneration: Long,
    val trustLevel: PeerTrustLevel,
    pendingStaticDhPublicKey: ByteArray? = null,
    val pendingGeneration: Long? = null,
) {
    private val _nodeId: ByteArray = nodeId.copyOf()
    private val _signingPublicKey: ByteArray = signingPublicKey.copyOf()
    private val _acceptedStaticDhPublicKey: ByteArray = acceptedStaticDhPublicKey.copyOf()
    private val _pendingStaticDhPublicKey: ByteArray? = pendingStaticDhPublicKey?.copyOf()

    val nodeId: ByteArray get() = _nodeId.copyOf()
    val signingPublicKey: ByteArray get() = _signingPublicKey.copyOf()
    val acceptedStaticDhPublicKey: ByteArray get() = _acceptedStaticDhPublicKey.copyOf()
    val pendingStaticDhPublicKey: ByteArray? get() = _pendingStaticDhPublicKey?.copyOf()

    val effectiveState: EffectivePeerTrustState
        get() = when {
            trustLevel == PeerTrustLevel.REVOKED -> EffectivePeerTrustState.REVOKED
            pendingGeneration != null -> EffectivePeerTrustState.KEY_CHANGED_QUARANTINED
            trustLevel == PeerTrustLevel.USER_VERIFIED -> EffectivePeerTrustState.ACTIVE_USER_VERIFIED
            else -> EffectivePeerTrustState.ACTIVE_TOFU
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PeerIdentityRecord) return false
        return _nodeId.contentEquals(other._nodeId) &&
               _signingPublicKey.contentEquals(other._signingPublicKey) &&
               _acceptedStaticDhPublicKey.contentEquals(other._acceptedStaticDhPublicKey) &&
               acceptedGeneration == other.acceptedGeneration &&
               trustLevel == other.trustLevel &&
               ((_pendingStaticDhPublicKey == null && other._pendingStaticDhPublicKey == null) ||
                (_pendingStaticDhPublicKey != null && other._pendingStaticDhPublicKey != null &&
                 _pendingStaticDhPublicKey.contentEquals(other._pendingStaticDhPublicKey))) &&
               pendingGeneration == other.pendingGeneration
    }

    override fun hashCode(): Int {
        var result = _nodeId.contentHashCode()
        result = 31 * result + _signingPublicKey.contentHashCode()
        result = 31 * result + _acceptedStaticDhPublicKey.contentHashCode()
        result = 31 * result + acceptedGeneration.hashCode()
        result = 31 * result + trustLevel.hashCode()
        result = 31 * result + (_pendingStaticDhPublicKey?.contentHashCode() ?: 0)
        result = 31 * result + (pendingGeneration?.hashCode() ?: 0)
        return result
    }
}

/**
 * Corruption reasons for durable [PeerIdentityRecord] validation (ADR-003, Phase C8.2A).
 */
internal enum class PeerRecordCorruptionReason {
    InvalidNodeIdLength,
    InvalidSigningKeyLength,
    InvalidAcceptedStaticKeyLength,
    AcceptedGenerationOutOfRange,
    NodeIdSigningKeyMismatch,
    PendingCouplingViolation,
    InvalidPendingStaticKeyLength,
    PendingGenerationOutOfRange,
    PendingNotNewer,
    PendingStaticEqualsAccepted,
    RevokedWithPending
}

/**
 * Validation result taxonomy for durable [PeerIdentityRecord] (ADR-003, Phase C8.2A).
 */
internal sealed class PeerIdentityRecordValidationResult {
    object Valid : PeerIdentityRecordValidationResult()
    data class Corrupt(val reason: PeerRecordCorruptionReason) : PeerIdentityRecordValidationResult()
}

/**
 * Pure validator enforcing the 11 durable invariants of [PeerIdentityRecord] (ADR-003, Phase C8.2A).
 *
 * Proves structural dimensions, domain ranges, cryptographic node_id derivation, and field coupling.
 * Does NOT prove repository provenance or out-of-band user verification.
 */
internal object PeerIdentityRecordValidator {
    const val NODE_ID_LENGTH = 16
    const val SIGNING_KEY_LENGTH = 32
    const val STATIC_KEY_LENGTH = 32
    const val MAX_UINT32 = 0xFFFFFFFFL

    fun validate(record: PeerIdentityRecord): PeerIdentityRecordValidationResult {
        // R1: nodeId exactly 16 bytes
        val nodeBytes = record.nodeId
        if (nodeBytes.size != NODE_ID_LENGTH) {
            return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.InvalidNodeIdLength)
        }

        // R2: signingPublicKey exactly 32 bytes
        val signBytes = record.signingPublicKey
        if (signBytes.size != SIGNING_KEY_LENGTH) {
            return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.InvalidSigningKeyLength)
        }

        // R3: acceptedStaticDhPublicKey exactly 32 bytes
        val accStatic = record.acceptedStaticDhPublicKey
        if (accStatic.size != STATIC_KEY_LENGTH) {
            return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.InvalidAcceptedStaticKeyLength)
        }

        // R4: acceptedGeneration within uint32 domain
        if (record.acceptedGeneration !in 0L..MAX_UINT32) {
            return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.AcceptedGenerationOutOfRange)
        }

        // R5: nodeId == BLAKE2s-128(signingPublicKey)
        val expectedNodeId = IdentityBindingV1.deriveNodeId(signBytes)
        if (!nodeBytes.contentEquals(expectedNodeId)) {
            return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.NodeIdSigningKeyMismatch)
        }

        // R6: pendingStaticDhPublicKey and pendingGeneration are BOTH null or BOTH non-null
        val pendStatic = record.pendingStaticDhPublicKey
        val pendGen = record.pendingGeneration
        if ((pendStatic == null) != (pendGen == null)) {
            return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.PendingCouplingViolation)
        }

        // Pending field invariants
        if (pendStatic != null && pendGen != null) {
            // R11: if trustLevel == REVOKED, pending fields MUST both be null
            if (record.trustLevel == PeerTrustLevel.REVOKED) {
                return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.RevokedWithPending)
            }

            // R7: pending static exactly 32 bytes
            if (pendStatic.size != STATIC_KEY_LENGTH) {
                return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.InvalidPendingStaticKeyLength)
            }

            // R9: pendingGeneration <= MAX_UINT32 and >= 0
            if (pendGen !in 0L..MAX_UINT32) {
                return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.PendingGenerationOutOfRange)
            }

            // R8: pendingGeneration > acceptedGeneration
            if (pendGen <= record.acceptedGeneration) {
                return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.PendingNotNewer)
            }

            // R10: pendingStaticDhPublicKey != acceptedStaticDhPublicKey
            if (pendStatic.contentEquals(accStatic)) {
                return PeerIdentityRecordValidationResult.Corrupt(PeerRecordCorruptionReason.PendingStaticEqualsAccepted)
            }
        }

        return PeerIdentityRecordValidationResult.Valid
    }
}

/**
 * Rejection reason taxonomy for peer trust evaluation (ADR-003, Phase C8.2A).
 */
internal enum class PeerTrustRejectReason {
    Rollback,
    SameGenerationConflict,
    PendingGenerationConflict,
    StaleRelativeToPending,
    NoncanonicalGenerationAdvance,
    NodeIdSigningKeyCollision,
    Revoked
}

/**
 * Pure decision plan emitted by [PeerTrustEngine] (ADR-003, Phase C8.2A).
 */
internal sealed class TrustPlan {
    object AcceptExisting : TrustPlan()
    object InsertFirstSeen : TrustPlan()
    object SetInitialPendingCandidate : TrustPlan()
    object AdvancePendingCandidate : TrustPlan()
    object KeepQuarantined : TrustPlan()
    data class Reject(val reason: PeerTrustRejectReason) : TrustPlan()
}

/**
 * Read-only view of a verified active peer identity (ADR-003, Phase C8.2A).
 */
class VerifiedPeerIdentity private constructor(
    nodeId: ByteArray,
    signingPublicKey: ByteArray,
    acceptedStaticDhPublicKey: ByteArray,
    val acceptedGeneration: Long,
    val trustLevel: PeerTrustLevel,
) {
    private val _nodeId: ByteArray = nodeId.copyOf()
    private val _signingPublicKey: ByteArray = signingPublicKey.copyOf()
    private val _acceptedStaticDhPublicKey: ByteArray = acceptedStaticDhPublicKey.copyOf()

    val nodeId: ByteArray get() = _nodeId.copyOf()
    val signingPublicKey: ByteArray get() = _signingPublicKey.copyOf()
    val acceptedStaticDhPublicKey: ByteArray get() = _acceptedStaticDhPublicKey.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is VerifiedPeerIdentity) return false
        return _nodeId.contentEquals(other._nodeId) &&
               _signingPublicKey.contentEquals(other._signingPublicKey) &&
               _acceptedStaticDhPublicKey.contentEquals(other._acceptedStaticDhPublicKey) &&
               acceptedGeneration == other.acceptedGeneration &&
               trustLevel == other.trustLevel
    }

    override fun hashCode(): Int {
        var result = _nodeId.contentHashCode()
        result = 31 * result + _signingPublicKey.contentHashCode()
        result = 31 * result + _acceptedStaticDhPublicKey.contentHashCode()
        result = 31 * result + acceptedGeneration.hashCode()
        result = 31 * result + trustLevel.hashCode()
        return result
    }

    companion object {
        internal fun fromRecord(record: PeerIdentityRecord): VerifiedPeerIdentity? {
            if (PeerIdentityRecordValidator.validate(record) !is PeerIdentityRecordValidationResult.Valid) {
                return null
            }
            if (record.trustLevel == PeerTrustLevel.REVOKED || record.pendingGeneration != null) {
                return null
            }
            return VerifiedPeerIdentity(
                nodeId = record.nodeId,
                signingPublicKey = record.signingPublicKey,
                acceptedStaticDhPublicKey = record.acceptedStaticDhPublicKey,
                acceptedGeneration = record.acceptedGeneration,
                trustLevel = record.trustLevel
            )
        }
    }
}

/**
 * Read-only view of a quarantined peer identity with a pending rotation candidate (ADR-003, Phase C8.2A).
 */
class PendingPeerIdentity private constructor(
    nodeId: ByteArray,
    signingPublicKey: ByteArray,
    acceptedStaticDhPublicKey: ByteArray,
    val acceptedGeneration: Long,
    val trustLevel: PeerTrustLevel,
    pendingStaticDhPublicKey: ByteArray,
    val pendingGeneration: Long,
) {
    private val _nodeId: ByteArray = nodeId.copyOf()
    private val _signingPublicKey: ByteArray = signingPublicKey.copyOf()
    private val _acceptedStaticDhPublicKey: ByteArray = acceptedStaticDhPublicKey.copyOf()
    private val _pendingStaticDhPublicKey: ByteArray = pendingStaticDhPublicKey.copyOf()

    val nodeId: ByteArray get() = _nodeId.copyOf()
    val signingPublicKey: ByteArray get() = _signingPublicKey.copyOf()
    val acceptedStaticDhPublicKey: ByteArray get() = _acceptedStaticDhPublicKey.copyOf()
    val pendingStaticDhPublicKey: ByteArray get() = _pendingStaticDhPublicKey.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PendingPeerIdentity) return false
        return _nodeId.contentEquals(other._nodeId) &&
               _signingPublicKey.contentEquals(other._signingPublicKey) &&
               _acceptedStaticDhPublicKey.contentEquals(other._acceptedStaticDhPublicKey) &&
               acceptedGeneration == other.acceptedGeneration &&
               trustLevel == other.trustLevel &&
               _pendingStaticDhPublicKey.contentEquals(other._pendingStaticDhPublicKey) &&
               pendingGeneration == other.pendingGeneration
    }

    override fun hashCode(): Int {
        var result = _nodeId.contentHashCode()
        result = 31 * result + _signingPublicKey.contentHashCode()
        result = 31 * result + _acceptedStaticDhPublicKey.contentHashCode()
        result = 31 * result + acceptedGeneration.hashCode()
        result = 31 * result + trustLevel.hashCode()
        result = 31 * result + _pendingStaticDhPublicKey.contentHashCode()
        result = 31 * result + pendingGeneration.hashCode()
        return result
    }

    companion object {
        internal fun fromRecord(record: PeerIdentityRecord): PendingPeerIdentity? {
            if (PeerIdentityRecordValidator.validate(record) !is PeerIdentityRecordValidationResult.Valid) {
                return null
            }
            if (record.trustLevel == PeerTrustLevel.REVOKED || record.pendingGeneration == null || record.pendingStaticDhPublicKey == null) {
                return null
            }
            return PendingPeerIdentity(
                nodeId = record.nodeId,
                signingPublicKey = record.signingPublicKey,
                acceptedStaticDhPublicKey = record.acceptedStaticDhPublicKey,
                acceptedGeneration = record.acceptedGeneration,
                trustLevel = record.trustLevel,
                pendingStaticDhPublicKey = record.pendingStaticDhPublicKey!!,
                pendingGeneration = record.pendingGeneration
            )
        }
    }
}
